### PR TITLE
feat: platform abstraction

### PR DESCRIPTION
**Pull request by KodingDev**
Perfectly happy with how it is implemented in Nucleus currently. We want to be able to have a single common API which can be used to write plugins / extensions in & reuse as much as possible in commons. We can probably mitigate some of these because Adventure has evolved a bit?

**TLDR**: I want to be able to keep stuff like optimized scoreboard logic in the common so I don’t need to re-implement it between platforms, or at least minimally so.

## TODO

- [x] Basic POC impl
- [x] Add some standard entity methods (mainly player interactions)
- [x] Audience implementations
- [ ] Platforms
  - [x] Fabric
  - [x] Paper
  - [ ] Velocity

### CODE REVIEW

File: core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/sided/PlatformServer.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * @author Koding
+ */
+interface PlatformServer {
+    companion object : PlatformBinding<PlatformServer>()
+
+    /**
+     * A list of all worlds on the server.
+     */
+    val worlds: List<PlatformWorld> get() = emptyList()
```
Comment (caoimhebyrne) [inline core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/sided/PlatformServer.kt]: Does the order of worlds matter? If not, why don't we use a `Set`?

Either that, or, would it not be better to use a `Map<String, PlatformWorld>`, with the key being the world's name/identifier? 

Right now, if I wanted to get the end world, right now I'd have to cycle through through all of the worlds until I find the end world, presumably through `firstOrNull`.

Comment (KodingDev) [reply core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/sided/PlatformServer.kt]: Check through the class again quickly, does that resolve the issues or should I still adjust?

File: core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/entity/PlatformEntity.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import org.joml.Vector2f
+import org.joml.Vector3d
+import java.util.UUID
+
+/**
+ * Abstraction for an entity, representing a living or non-living object in the world.
+ * This should assume the entity is in the world in most cases, however there is a
+ * field to check if the entity is valid.
+ *
+ * TODO: Implement
```
Comment (KodingDev) [inline core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/entity/PlatformEntity.kt]: Remove TODO

File: core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/entity/PlatformEntity.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+     * == Identified ==
+     */
+    override fun identity(): Identity = Identity.identity(uuid)
+
+    class Type {
+        companion object {
+            @JvmStatic
+            val PLAYER = Key.key("minecraft", "player")
+        }
+    }
```
Comment (KodingDev) [inline core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/entity/PlatformEntity.kt]: Extract out to another class / enum?

File: core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/sided/PlatformServer.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * A list of all players on the server.
+     */
+    val players: List<PlatformPlayer> get() = emptyList()
```
Comment (KodingDev) [inline core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/sided/PlatformServer.kt]: Make these throw an exception

File: core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/world/PlatformWorld.kt [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * The maximum y coordinate of the world. Defaults to 255.
+     */
+    val maxY: Int get() = 255
+
+    /**
+     * A list of all the entities in the world.
+     */
+    val entities: List<PlatformEntity> get() = emptyList()
```
Comment (KodingDev) [inline core/common/src/main/kotlin/dev/koding/catalyst/core/common/api/platform/world/PlatformWorld.kt]: Throw unsupported.

File: core/fabric/build.gradle.kts [hunk 6]
```diff
@@ -47,13 +46,20 @@ dependencies {
...
     // Mods
     "modRuntimeOnly"("me.djtheredstoner:DevAuth-fabric:1.1.0")
     "modRuntimeOnly"("maven.modrinth:modmenu:4.1.1")
 }
 
 configure<LoomGradleExtensionAPI> {
-    silentMojangMappingsLicense()
+    runConfigs {
+        all {
+            isIdeConfigGenerated = true
```
Comment (KodingDev) [inline core/fabric/build.gradle.kts:60]: Describe why this is needed.

File: core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/entity/player/FabricPlatformEntity.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    override val type = Key.key(entityKey.namespace, entityKey.path)
+
+    override val uuid: UUID = ref.uuid
+    override val id: Int = ref.id
+
+    override val position: Vector3d
+        get() = Vector3d(ref.position().x, ref.position().y, ref.position().z)
+
+    override val look: Vector2f
+        get() = Vector2f(ref.yRot, ref.xRot)
```
Comment (KodingDev) [inline core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/entity/player/FabricPlatformEntity.kt:46]: Inline getters

File: core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/entity/player/FabricPlatformEntity.kt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    override val world: PlatformWorld get() = ref.level.wrap()
+
+    override fun teleport(position: Vector3d, look: Vector2f?, world: PlatformWorld?): Boolean {
+        // TODO: Use world and look
```
Comment (KodingDev) [inline core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/entity/player/FabricPlatformEntity.kt:51]: Complete this TODO

File: core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/entity/player/FabricPlatformEntity.kt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+fun PlatformEntity.unwrap(): Entity = (this as FabricPlatformEntity).ref
+
+/**
+ * Performs smart wrapping of an entity into a Catalyst entity of the correct type.
+ */
+fun Entity.wrap(): PlatformEntity = when (this) {
+    is Player -> wrap()
+    is LivingEntity -> wrap()
+    else -> wrapRaw()
+}
```
Comment (KodingDev) [inline core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/entity/player/FabricPlatformEntity.kt]: Rename all other extensions to wrap and change to wrapSmart()

File: core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/world/FabricPlatformBlock.kt [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import org.joml.Vector3i
+import net.minecraft.core.BlockPos as MCBlockPos
+
+class FabricPlatformBlock(
+    private val ref: Block,
+    override val position: BlockPos,
+    override val chunk: PlatformChunk
+) : PlatformBlock {
+    override val type: BlockMetadata
+        get() = ref.defaultBlockState().wrap()
```
Comment (KodingDev) [inline core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/world/FabricPlatformBlock.kt:37]: Inline get()

File: core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/world/FabricPlatformChunk.kt [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    override fun setBlock(x: Int, y: Int, z: Int, block: BlockMetadata) {
+//        val position = BlockPos(x, y, z)
+//        val state = Block.stateById(block.type)
+//
+//        ref.setBlockState(position)
+
+        throw NotImplementedError()
+    }
```
Comment (KodingDev) [inline core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/world/FabricPlatformChunk.kt]: Remove this

File: core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/world/FabricPlatformChunk.kt [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+class FabricPlatformChunk(
+    private val ref: LevelChunk,
+    override val world: PlatformWorld
+) : PlatformChunk {
+    override val x: Int
+        get() = ref.pos.x
+
+    override val z: Int
+        get() = ref.pos.z
```
Comment (KodingDev) [inline core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/world/FabricPlatformChunk.kt:39]: Inline get()

File: core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/world/FabricPlatformWorld.kt [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        get() = ref.minBuildHeight
+
+    override val maxY: Int
+        get() = ref.maxBuildHeight
+
+    override val players: List<PlatformPlayer>
+        get() = ref.players().map { it.wrap() }
+
+    override val entities: List<PlatformEntity>
+        get() = ref.allEntities.map { it.wrap() }
```
Comment (KodingDev) [inline core/fabric/src/main/kotlin/dev/koding/catalyst/core/fabric/api/platform/world/FabricPlatformWorld.kt]: Inline get()

File: core/paper/src/main/kotlin/dev/koding/catalyst/core/paper/CatalystPlugin.kt [hunk 14]
```diff
@@ -32,5 +39,31 @@ class CatalystPlugin : PaperLoader({
...
+                chunk.setBlock(0, i, 0, BlockMetadata.AIR)
+            }
+        }
+
+        SchedulersImpl.sync.schedule(0L, 500L) {
+            PlatformServerImpl.players.forEach {
+                it.teleport(it.position.add(0.0, 0.5, 0.0))
+                it.sendMessage(Component.text("lol"))
+            }
+        }
```
Comment (KodingDev) [inline core/paper/src/main/kotlin/dev/koding/catalyst/core/paper/CatalystPlugin.kt]: Remove this

File: core/paper/src/main/kotlin/dev/koding/catalyst/core/paper/api/platform/entity/PaperPlatformEntity.kt [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Performs smart wrapping of a Bukkit entity into a Catalyst entity
+ * of the correct type.
+ */
+fun Entity.wrap(): PlatformEntity = when (this) {
+    is Player -> wrap()
+    is LivingEntity -> wrap()
+    else -> wrapRaw()
+}
```
Comment (KodingDev) [inline core/paper/src/main/kotlin/dev/koding/catalyst/core/paper/api/platform/entity/PaperPlatformEntity.kt:56]: wrap -> smartWrap(), remove wrapRaw()

**Review by KodingDev — COMMENTED** [review]
Few fixes which are needed prior to merging

**Pull request merged**